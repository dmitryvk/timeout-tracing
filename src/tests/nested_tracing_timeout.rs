use std::time::Duration;

use serial_test::serial;
use tokio::time::sleep;
use tracing::instrument;

use crate::{
    TimeoutElapsed,
    tests::{insta_trace_filters, run_with_tracing},
    timeout,
    trace::CaptureSpanAndStackTrace,
};

#[tokio::test]
#[serial]
async fn with_timeouts() {
    let result = run_with_tracing(Duration::from_millis(100), do_f()).await;

    assert!(matches!(result, Err(TimeoutElapsed { .. })));
    let mut err = result.err().unwrap();
    err.active_traces
        .sort_by_cached_key(|trace| trace.span_trace.to_string());
    insta::with_settings!({
        filters => insta_trace_filters()
    }, {
        insta::assert_debug_snapshot!(err);
        insta::assert_snapshot!(err);
    });
}

#[instrument]
async fn do_f() {
    _ = timeout(Duration::from_secs(1), CaptureSpanAndStackTrace, do_g()).await;
}

#[instrument]
async fn do_g() {
    sleep(Duration::from_secs(2)).await;
}
