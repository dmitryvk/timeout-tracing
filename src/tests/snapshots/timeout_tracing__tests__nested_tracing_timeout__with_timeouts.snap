---
source: src/tests/nested_tracing_timeout.rs
expression: err
---
TimeoutElapsed {
    active_traces: [
        StackAndSpanTrace {
            stack_trace: Backtrace [
                { fn: "timeout_tracing::trace::StackAndSpanTrace::capture", file: "./src/trace.rs", line: [NNN] },
                { fn: "<timeout_tracing::trace::DefaultTraceCapturer as timeout_tracing::trace::CaptureTrace>::capture", file: "./src/trace.rs", line: [NNN] },
                { fn: "timeout_tracing::waker::TracingTimeoutWaker<C>::clone", file: "./src/waker.rs", line: [NNN] },
                { fn: "timeout_tracing::waker::TracingTimeoutWaker<C>::raw_clone", file: "./src/waker.rs", line: [NNN] },
                { fn: "<&core::task::wake::Waker as tokio::sync::task::atomic_waker::WakerRef>::into_waker", file: "[crates]/tokio-[ver]/src/sync/task/atomic_waker.rs", line: [NNN] },
                { fn: "tokio::sync::task::atomic_waker::AtomicWaker::do_register::{{closure}}", file: "[crates]/tokio-[ver]/src/sync/task/atomic_waker.rs", line: [NNN] },
                { fn: "__rust_try" },
                { fn: "tokio::sync::task::atomic_waker::AtomicWaker::do_register::catch_unwind", file: "[crates]/tokio-[ver]/src/sync/task/atomic_waker.rs", line: [NNN] },
                { fn: "tokio::sync::task::atomic_waker::AtomicWaker::do_register", file: "[crates]/tokio-[ver]/src/sync/task/atomic_waker.rs", line: [NNN] },
                { fn: "tokio::sync::task::atomic_waker::AtomicWaker::register_by_ref", file: "[crates]/tokio-[ver]/src/sync/task/atomic_waker.rs", line: [NNN] },
                { fn: "tokio::runtime::time::entry::StateCell::poll", file: "[crates]/tokio-[ver]/src/runtime/time/entry.rs", line: [NNN] },
                { fn: "tokio::runtime::time::entry::TimerEntry::poll_elapsed", file: "[crates]/tokio-[ver]/src/runtime/time/entry.rs", line: [NNN] },
                { fn: "tokio::time::sleep::Sleep::poll_elapsed", file: "[crates]/tokio-[ver]/src/time/sleep.rs", line: [NNN] },
                { fn: "<tokio::time::sleep::Sleep as core::future::future::Future>::poll", file: "[crates]/tokio-[ver]/src/time/sleep.rs", line: [NNN] },
                { fn: "<timeout_tracing::TimeoutFuture<C,Fut> as core::future::future::Future>::poll", file: "./src/lib.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::nested_tracing_timeout::do_f::{{closure}}::{{closure}}", file: "./src/tests/nested_tracing_timeout.rs", line: [NNN] },
                { fn: "<tracing::instrument::Instrumented<T> as core::future::future::Future>::poll", file: "[crates]/tracing-[ver]/src/instrument.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::nested_tracing_timeout::do_f::{{closure}}", file: "./src/tests/nested_tracing_timeout.rs", line: [NNN] },
                { fn: "<timeout_tracing::TimeoutFuture<C,Fut> as core::future::future::Future>::poll", file: "./src/lib.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::run_with_tracing::{{closure}}", file: "./src/tests/mod.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::nested_tracing_timeout::with_timeouts::{{closure}}::{{closure}}", file: "./src/tests/nested_tracing_timeout.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}::{{closure}}", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::task::coop::with_budget", file: "[crates]/tokio-[ver]/src/task/coop/mod.rs", line: [NNN] },
                { fn: "tokio::task::coop::budget", file: "[crates]/tokio-[ver]/src/task/coop/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::Context::enter", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CoreGuard::enter::{{closure}}", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::context::scoped::Scoped<T>::set", file: "[crates]/tokio-[ver]/src/runtime/context/scoped.rs", line: [NNN] },
                { fn: "tokio::runtime::context::set_scheduler::{{closure}}", file: "[crates]/tokio-[ver]/src/runtime/context.rs", line: [NNN] },
                { fn: "tokio::runtime::context::set_scheduler", file: "[crates]/tokio-[ver]/src/runtime/context.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CoreGuard::enter", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CoreGuard::block_on", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CurrentThread::block_on::{{closure}}", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::context::runtime::enter_runtime", file: "[crates]/tokio-[ver]/src/runtime/context/runtime.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CurrentThread::block_on", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::runtime::Runtime::block_on_inner", file: "[crates]/tokio-[ver]/src/runtime/runtime.rs", line: [NNN] },
                { fn: "tokio::runtime::runtime::Runtime::block_on", file: "[crates]/tokio-[ver]/src/runtime/runtime.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::nested_tracing_timeout::with_timeouts::{{closure}}", file: "./src/tests/nested_tracing_timeout.rs", line: [NNN] },
                { fn: "serial_test::serial_code_lock::local_serial_core", file: "[crates]/serial_test-[ver]/src/serial_code_lock.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::nested_tracing_timeout::with_timeouts", file: "./src/tests/nested_tracing_timeout.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::nested_tracing_timeout::with_timeouts::{{closure}}", file: "./src/tests/nested_tracing_timeout.rs", line: [NNN] },
            ],
            span_trace: SpanTrace [
                { target: "timeout_tracing", name: "deadline", file: "src/lib.rs", line: [NNN] },
                { target: "timeout_tracing::tests::nested_tracing_timeout", name: "do_f", file: "src/tests/nested_tracing_timeout.rs", line: [NNN] },
            ],
        },
        StackAndSpanTrace {
            stack_trace: Backtrace [
                { fn: "timeout_tracing::trace::StackAndSpanTrace::capture", file: "./src/trace.rs", line: [NNN] },
                { fn: "<timeout_tracing::trace::DefaultTraceCapturer as timeout_tracing::trace::CaptureTrace>::capture", file: "./src/trace.rs", line: [NNN] },
                { fn: "timeout_tracing::waker::TracingTimeoutWaker<C>::clone", file: "./src/waker.rs", line: [NNN] },
                { fn: "timeout_tracing::waker::TracingTimeoutWaker<C>::raw_clone", file: "./src/waker.rs", line: [NNN] },
                { fn: "<&core::task::wake::Waker as tokio::sync::task::atomic_waker::WakerRef>::into_waker", file: "[crates]/tokio-[ver]/src/sync/task/atomic_waker.rs", line: [NNN] },
                { fn: "tokio::sync::task::atomic_waker::AtomicWaker::do_register::{{closure}}", file: "[crates]/tokio-[ver]/src/sync/task/atomic_waker.rs", line: [NNN] },
                { fn: "__rust_try" },
                { fn: "tokio::sync::task::atomic_waker::AtomicWaker::do_register::catch_unwind", file: "[crates]/tokio-[ver]/src/sync/task/atomic_waker.rs", line: [NNN] },
                { fn: "tokio::sync::task::atomic_waker::AtomicWaker::do_register", file: "[crates]/tokio-[ver]/src/sync/task/atomic_waker.rs", line: [NNN] },
                { fn: "tokio::sync::task::atomic_waker::AtomicWaker::register_by_ref", file: "[crates]/tokio-[ver]/src/sync/task/atomic_waker.rs", line: [NNN] },
                { fn: "tokio::runtime::time::entry::StateCell::poll", file: "[crates]/tokio-[ver]/src/runtime/time/entry.rs", line: [NNN] },
                { fn: "tokio::runtime::time::entry::TimerEntry::poll_elapsed", file: "[crates]/tokio-[ver]/src/runtime/time/entry.rs", line: [NNN] },
                { fn: "tokio::time::sleep::Sleep::poll_elapsed", file: "[crates]/tokio-[ver]/src/time/sleep.rs", line: [NNN] },
                { fn: "<tokio::time::sleep::Sleep as core::future::future::Future>::poll", file: "[crates]/tokio-[ver]/src/time/sleep.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::nested_tracing_timeout::do_g::{{closure}}::{{closure}}", file: "./src/tests/nested_tracing_timeout.rs", line: [NNN] },
                { fn: "<tracing::instrument::Instrumented<T> as core::future::future::Future>::poll", file: "[crates]/tracing-[ver]/src/instrument.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::nested_tracing_timeout::do_g::{{closure}}", file: "./src/tests/nested_tracing_timeout.rs", line: [NNN] },
                { fn: "<timeout_tracing::TimeoutFuture<C,Fut> as core::future::future::Future>::poll", file: "./src/lib.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::nested_tracing_timeout::do_f::{{closure}}::{{closure}}", file: "./src/tests/nested_tracing_timeout.rs", line: [NNN] },
                { fn: "<tracing::instrument::Instrumented<T> as core::future::future::Future>::poll", file: "[crates]/tracing-[ver]/src/instrument.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::nested_tracing_timeout::do_f::{{closure}}", file: "./src/tests/nested_tracing_timeout.rs", line: [NNN] },
                { fn: "<timeout_tracing::TimeoutFuture<C,Fut> as core::future::future::Future>::poll", file: "./src/lib.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::run_with_tracing::{{closure}}", file: "./src/tests/mod.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::nested_tracing_timeout::with_timeouts::{{closure}}::{{closure}}", file: "./src/tests/nested_tracing_timeout.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}::{{closure}}", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::task::coop::with_budget", file: "[crates]/tokio-[ver]/src/task/coop/mod.rs", line: [NNN] },
                { fn: "tokio::task::coop::budget", file: "[crates]/tokio-[ver]/src/task/coop/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::Context::enter", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CoreGuard::enter::{{closure}}", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::context::scoped::Scoped<T>::set", file: "[crates]/tokio-[ver]/src/runtime/context/scoped.rs", line: [NNN] },
                { fn: "tokio::runtime::context::set_scheduler::{{closure}}", file: "[crates]/tokio-[ver]/src/runtime/context.rs", line: [NNN] },
                { fn: "tokio::runtime::context::set_scheduler", file: "[crates]/tokio-[ver]/src/runtime/context.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CoreGuard::enter", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CoreGuard::block_on", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CurrentThread::block_on::{{closure}}", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::context::runtime::enter_runtime", file: "[crates]/tokio-[ver]/src/runtime/context/runtime.rs", line: [NNN] },
                { fn: "tokio::runtime::scheduler::current_thread::CurrentThread::block_on", file: "[crates]/tokio-[ver]/src/runtime/scheduler/current_thread/mod.rs", line: [NNN] },
                { fn: "tokio::runtime::runtime::Runtime::block_on_inner", file: "[crates]/tokio-[ver]/src/runtime/runtime.rs", line: [NNN] },
                { fn: "tokio::runtime::runtime::Runtime::block_on", file: "[crates]/tokio-[ver]/src/runtime/runtime.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::nested_tracing_timeout::with_timeouts::{{closure}}", file: "./src/tests/nested_tracing_timeout.rs", line: [NNN] },
                { fn: "serial_test::serial_code_lock::local_serial_core", file: "[crates]/serial_test-[ver]/src/serial_code_lock.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::nested_tracing_timeout::with_timeouts", file: "./src/tests/nested_tracing_timeout.rs", line: [NNN] },
                { fn: "timeout_tracing::tests::nested_tracing_timeout::with_timeouts::{{closure}}", file: "./src/tests/nested_tracing_timeout.rs", line: [NNN] },
            ],
            span_trace: SpanTrace [
                { target: "timeout_tracing::tests::nested_tracing_timeout", name: "do_g", file: "src/tests/nested_tracing_timeout.rs", line: [NNN] },
                { target: "timeout_tracing::tests::nested_tracing_timeout", name: "do_f", file: "src/tests/nested_tracing_timeout.rs", line: [NNN] },
            ],
        },
    ],
}
